src/utils.js

```diff
@@ -379,17 +380,82 @@ function findSchemaDefinition($ref, definitions = {}) {
   throw new Error(`Could not find a definition for ${$ref}.`);
 }
 
-export function retrieveSchema(schema, definitions = {}) {
-  // No $ref attribute found, returning the original schema.
-  if (!schema.hasOwnProperty("$ref")) {
+export function retrieveSchema(schema, definitions = {}, formData = {}) {
+  if (schema.hasOwnProperty("$ref")) {
+    // Retrieve the referenced schema definition.
+    const $refSchema = findSchemaDefinition(schema.$ref, definitions);
+    // Drop the $ref property of the source schema.
+    const { $ref, ...localSchema } = schema;
+    // Update referenced schema definition with local schema properties.
+    return retrieveSchema(
+      { ...$refSchema, ...localSchema },
+      definitions,
+      formData
+    );
+  } else if (schema.hasOwnProperty("dependencies")) {
+    // Drop the dependencies from the source schema.
+    let { dependencies, ...localSchema } = schema;
+    // Process dependencies updating the local schema properties as appropriate.
+    for (const key of Object.keys(dependencies)) {
+      // Skip this dependency if its trigger property is not present.
+      if (!formData[key]) {
+        // fixme: a falsey check may not be appropriate here, I'm not sure
+        continue;
+      }
+      const value = dependencies[key];
+      if (Array.isArray(value)) {
+        for (const property of value) {
+          if (!Array.isArray(localSchema.required)) {
+            localSchema.required = [];
+          }
+          if (localSchema.required.includes(property)) {
+            continue;
+          }
+          localSchema.required = localSchema.required.concat(property);
+        }
+      } else if (isObject(value)) {
+        localSchema = retrieveSchema(
+          mergeObjects(
+            localSchema,
+            retrieveSchema(value, definitions, formData),
+            true
+          )
+        );
+        if (Array.isArray(value.oneOf)) {
+          for (const subschema of value.oneOf) {
+            if (!subschema.properties) {
+              continue;
+            }
+            const {
+              [key]: conditionProperty,
+              ...dependentProperties
+            } = subschema.properties;
+            if (conditionProperty) {
+              const conditionSchema = {
+                type: "object",
+                properties: {
+                  [key]: conditionProperty,
+                },
+              };
+              const { errors } = jsonValidate(formData, conditionSchema);
+              if (errors.length === 0) {
+                const dependentSchema = { ...subschema };
```

Comment:
Correct. I am just trying to not mutate the incoming schema at all so I make copies as needed in the tree instead of recursively coping the entire tree (although I suppose the later could be done for readability/maintainability reasons although I'm not sure how to do that off the top of my head).